Src/Newtonsoft.Json/Utilities/CollectionUtils.cs

```diff
@@ -340,9 +344,27 @@ public static Array ToMultidimensionalArray(IList values, Type type, int rank)
             }
 
             Array multidimensionalArray = Array.CreateInstance(type, dimensions.ToArray());
-            CopyFromJaggedToMultidimensionalArray(values, multidimensionalArray, new int[0]);
+            CopyFromJaggedToMultidimensionalArray(values, multidimensionalArray, ArrayEmpty<int>());
 
             return multidimensionalArray;
         }
+
+        // 4.6 has Array.Empty<T> to return a cached empty array. Lacking that in other
+        // frameworks, Enumerable.Empty<T> happens to be implemented as a cached empty
+        // array in all versions (in .NET Core the same instance as Array.Empty<T>).
+        // This includes the internal Linq bridge for 2.0.
+        // Since this method is simple and only 11 bytes long in a release build it's
+        // pretty much guaranteed to be inlined, giving us fast access of that cached
+        // array. With 4.5 and up we use AggressiveInlining just to be sure, so it's
+        // effectively the same as calling Array.Empty<T> even when not available.
```

Comment:
The thing I worry about this is what happens if an implementation of `Enumerable.Empty<T>()` by someone at some point in time doesn't return an array? ArrayEmpty will start returning null.

We could put this code in a cctor and check if `Enumerable.Empty<T>()` is null or not, creating an empty array if it is, but that would require a new generic class to store the variable on.

Can you think of other options?